In [10]:
import pandas as pd

import numpy as np 

import matplotlib.pyplot as plt  

import seaborn as sns

In [11]:
file_path = "../data/raw/DDos.csv"

df = pd.read_csv(file_path)

df.head()

Destination Port   Flow Duration   Total Fwd Packets  \
0              54865               3                   2   
1              55054             109                   1   
2              55055              52                   1   
3              46236              34                   1   
4              54863               3                   2   

    Total Backward Packets  Total Length of Fwd Packets  \
0                        0                           12   
1                        1                            6   
2                        1                            6   
3                        1                            6   
4                        0                           12   

    Total Length of Bwd Packets   Fwd Packet Length Max  \
0                             0                       6   
1                             6                       6   
2                             6                       6   
3                             6                       6   
4                             0                       6   

    Fwd Packet Length Min   Fwd Packet Length Mean   Fwd Packet Length Std  \
0                       6                      6.0                     0.0   
1                       6                      6.0                     0.0   
2                       6                      6.0                     0.0   
3                       6                      6.0                     0.0   
4                       6                      6.0                     0.0   

   ...   min_seg_size_forward  Active Mean   Active Std   Active Max  \
0  ...                   20.0          0.0          0.0          0.0   
1  ...                   20.0          0.0          0.0          0.0   
2  ...                   20.0          0.0          0.0          0.0   
3  ...                   20.0          0.0          0.0          0.0   
4  ...                   20.0          0.0          0.0          0.0   

    Active Min  Idle Mean   Idle Std   Idle Max   Idle Min   Label  
0          0.0        0.0        0.0        0.0        0.0  BENIGN  
1          0.0        0.0        0.0        0.0        0.0  BENIGN  
2          0.0        0.0        0.0        0.0        0.0  BENIGN  
3          0.0        0.0        0.0        0.0        0.0  BENIGN  
4          0.0        0.0        0.0        0.0        0.0  BENIGN  

[5 rows x 79 columns]

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 66237 entries, 0 to 66236
Data columns (total 79 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0    Destination Port             66237 non-null  int64  
 1    Flow Duration                66237 non-null  int64  
 2    Total Fwd Packets            66237 non-null  int64  
 3    Total Backward Packets       66237 non-null  int64  
 4   Total Length of Fwd Packets   66237 non-null  int64  
 5    Total Length of Bwd Packets  66237 non-null  int64  
 6    Fwd Packet Length Max        66237 non-null  int64  
 7    Fwd Packet Length Min        66237 non-null  int64  
 8    Fwd Packet Length Mean       66237 non-null  float64
 9    Fwd Packet Length Std        66237 non-null  float64
 10  Bwd Packet Length Max         66237 non-null  int64  
 11   Bwd Packet Length Min        66237 non-null  int64  
 12   Bwd Packet Length Mean       66237 non-null  float64
 13   

#### Data preprocessing

In [13]:
df.isnull().sum()

 Destination Port              0
 Flow Duration                 0
 Total Fwd Packets             0
 Total Backward Packets        0
Total Length of Fwd Packets    0
                              ..
Idle Mean                      1
 Idle Std                      1
 Idle Max                      1
 Idle Min                      1
 Label                         1
Length: 79, dtype: int64

In [15]:
df.columns = df.columns.str.strip()

In [18]:
df.loc[:,'Label'].unique()

array(['BENIGN', 'DDoS', nan], dtype=object)

In [19]:
df['Label'].value_counts()


Label
DDoS      34952
BENIGN    31284
Name: count, dtype: int64

In [20]:
data_nona=df.dropna()

In [21]:
data_nona.info()

<class 'pandas.core.frame.DataFrame'>
Index: 66236 entries, 0 to 66235
Data columns (total 79 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   Destination Port             66236 non-null  int64  
 1   Flow Duration                66236 non-null  int64  
 2   Total Fwd Packets            66236 non-null  int64  
 3   Total Backward Packets       66236 non-null  int64  
 4   Total Length of Fwd Packets  66236 non-null  int64  
 5   Total Length of Bwd Packets  66236 non-null  int64  
 6   Fwd Packet Length Max        66236 non-null  int64  
 7   Fwd Packet Length Min        66236 non-null  int64  
 8   Fwd Packet Length Mean       66236 non-null  float64
 9   Fwd Packet Length Std        66236 non-null  float64
 10  Bwd Packet Length Max        66236 non-null  int64  
 11  Bwd Packet Length Min        66236 non-null  int64  
 12  Bwd Packet Length Mean       66236 non-null  float64
 13  Bwd Packet Length Std

In [22]:
data_nona['Label'].value_counts()

Label
DDoS      34952
BENIGN    31284
Name: count, dtype: int64

In [23]:
data_nona.loc[:,'Label'].unique()

array(['BENIGN', 'DDoS'], dtype=object)

In [24]:
data_nona.describe()

Destination Port  Flow Duration  Total Fwd Packets  \
count      66236.000000   6.623600e+04       66236.000000   
mean        6454.701794   1.388757e+07           5.180295   
std        16711.002247   2.815859e+07          16.873784   
min            0.000000   0.000000e+00           1.000000   
25%           80.000000   4.995450e+04           2.000000   
50%           80.000000   9.949430e+05           3.000000   
75%           80.000000   7.640073e+06           5.000000   
max        61538.000000   1.199983e+08        1681.000000   

       Total Backward Packets  Total Length of Fwd Packets  \
count            66236.000000                 66236.000000   
mean                 5.119225                   695.632677   
std                 23.797303                  3230.070242   
min                  0.000000                     0.000000   
25%                  1.000000                    26.000000   
50%                  4.000000                    30.000000   
75%                  5.000000                    64.000000   
max               2942.000000                120783.000000   

       Total Length of Bwd Packets  Fwd Packet Length Max  \
count                 6.623600e+04           66236.000000   
mean                  6.716764e+03             351.447068   
std                   4.222729e+04            1435.520764   
min                   0.000000e+00               0.000000   
25%                   0.000000e+00               6.000000   
50%                   2.020000e+02              20.000000   
75%                   1.160100e+04              38.000000   
max                   4.991419e+06           11680.000000   

       Fwd Packet Length Min  Fwd Packet Length Mean  Fwd Packet Length Std  \
count           66236.000000            66236.000000           66236.000000   
mean               23.466151              110.716876             135.440531   
std               136.609404              400.379518             613.426001   
min                 0.000000                0.000000               0.000000   
25%                 0.000000                6.000000               0.000000   
50%                 0.000000                8.666667               0.000000   
75%                 6.000000               35.000000              10.263203   
max              1472.000000             3867.000000            6692.644993   

       ...  act_data_pkt_fwd  min_seg_size_forward   Active Mean  \
count  ...      66236.000000          66236.000000  6.623600e+04   
mean   ...          3.447793             22.101455  1.585840e+05   
std    ...         13.108006              4.844533  8.489147e+05   
min    ...          0.000000              0.000000  0.000000e+00   
25%    ...          1.000000             20.000000  0.000000e+00   
50%    ...          2.000000             20.000000  0.000000e+00   
75%    ...          4.000000             20.000000  1.005000e+03   
max    ...       1680.000000             52.000000  1.000000e+08   

         Active Std    Active Max    Active Min     Idle Mean      Idle Std  \
count  6.623600e+04  6.623600e+04  6.623600e+04  6.623600e+04  6.623600e+04   
mean   2.020972e+04  1.949634e+05  1.470892e+05  8.153964e+06  2.602819e+06   
std    3.050515e+05  1.028144e+06  8.214214e+05  1.804617e+07  9.841559e+06   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
75%    0.000000e+00  1.005000e+03  1.005000e+03  7.410306e+06  0.000000e+00   
max    3.950000e+07  1.000000e+08  1.000000e+08  1.200000e+08  6.530000e+07   

           Idle Max      Idle Min  
count  6.623600e+04  6.623600e+04  
mean   9.997795e+06  6.297389e+06  
std    2.187645e+07  1.642701e+07  
min    0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  
50%    0.000000e+00  0.000000e+00  
75%    7.415575e+06  6.526832e+06  
max    1.200000e+08  1.

#### Explore the data

In [25]:
# Filter for DDoS attacks
ddos_df = data_nona[data_nona['Label'] == 'DDoS']


In [26]:
# Get basic statistics for DDoS data
print(ddos_df.describe())


       Destination Port  Flow Duration  Total Fwd Packets  \
count           34952.0   3.495200e+04       34952.000000   
mean               80.0   1.370784e+07           4.317550   
std                 0.0   2.561829e+07           1.834958   
min                80.0   0.000000e+00           1.000000   
25%                80.0   5.786868e+05           3.000000   
50%                80.0   1.701951e+06           4.000000   
75%                80.0   8.379818e+06           5.000000   
max                80.0   8.137270e+07          10.000000   

       Total Backward Packets  Total Length of Fwd Packets  \
count            34952.000000                 34952.000000   
mean                 3.345989                    31.113870   
std                  2.538891                    11.401108   
min                  0.000000                     6.000000   
25%                  0.000000                    26.000000   
50%                  4.000000                    26.000000   
75%             

1. Most of the DDoS attacks are targeting **port 80**, which is commonly used for HTTP traffic. This is typical because attackers often target web services.
2. The duration of DDoS attack flows varies greatly. A very high Flow Duration could indicate a long-lasting attack. For instance, an attack lasting 13 seconds could be a sustained attempt to overwhelm the server.
3. "Forward" means the direction from the sender (in this case,  the attacker) to the receiver (the target of the DDoS attack).
- *Ping Flood* Attack. The attacker attempts to overwhelm a targeted device with ICMP echo-request packets Each ping might be around 28-32 bytes.This fits our mean of 31.11 bytes.The server gets overwhelmed trying to respond to all these small messages.
- *SYN Flood* Attack.A type of network-based denial-of-service (DoS) attack in which an attacker overwhelms a target system by sending a large number of SYN (synchronize) requests to the target's server. 
- *Low-and-Slow* Attack. This type of attack uses seemingly normal traffic but in a way that exhausts server resources.Packet sizes might vary more, explaining our standard deviation of 11.40 bytes.The attack is harder to detect because it mimics normal traffic patterns.
- *Amplification Attack* This type of attack involves larger packets, but our statistics suggest a different approach of the attacks with a maximum of 68 bytes which is still small.
4. 
